### PR TITLE
Rollback if release stuck in pending

### PR DESCRIPTION
**Pull request by tomahawk28**
Toward https://github.com/giantswarm/giantswarm/issues/12245

See https://github.com/giantswarm/giantswarm/issues/12245#issuecomment-667996388 for the dev flow on this change.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -9,7 +9,7 @@ require (
 	github.com/giantswarm/appcatalog v0.2.7
 	github.com/giantswarm/backoff v0.2.0
 	github.com/giantswarm/exporterkit v0.2.0
-	github.com/giantswarm/helmclient v1.0.5
+	github.com/giantswarm/helmclient v1.0.6-0.20200731115526-936ce495d21e
```
Comment (tomahawk28) [inline go.mod:12]: TODO: release new helmclient before merge

File: service/controller/chart/resource/release/update.go [hunk 2]
```diff
@@ -211,8 +220,15 @@ func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desir
 	// The release is still being updated so we don't update and check again
 	// in the next reconciliation loop.
 	if isReleaseInTransitionState(currentReleaseState) {
-		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("the %#q release is in status %#q and cannot be updated", desiredReleaseState.Name, currentReleaseState.Status))
-		return nil, nil
+		upgradeForce := key.HasForceUpgradeAnnotation(cr)
+		// Only perform a rollback in case of upgrade force is enabled.
+		// This is to consider critical app's service level and stateful apps.
+		if upgradeForce {
```
Comment (tomahawk28) [inline service/controller/chart/resource/release/update.go:226]: Only perform rollbacks in case of `upgradeForce` is enabled.

**Comment by tomahawk28** [comment]
Tested on `ginger`, `godsmack` with `pending-install`, `pending-upgrade` release.

File: main.go [hunk 3]
```diff
@@ -108,6 +108,7 @@ func mainWithError() error {
 
 	daemonCommand.PersistentFlags().String(f.Service.Helm.HTTP.ClientTimeout, "5s", "HTTP timeout for pulling chart tarballs.")
 	daemonCommand.PersistentFlags().String(f.Service.Helm.Kubernetes.WaitTimeout, "10s", "Wait timeout when calling the Kubernetes API.")
+	daemonCommand.PersistentFlags().Int(f.Service.Helm.MaxRollback, 3, "the maximum number of rollback retries for pending apps.")
```
Comment (rossf7) [inline main.go:111]:
```suggestion
	daemonCommand.PersistentFlags().Int(f.Service.Helm.MaxRollback, 3, "the maximum number of rollback attempts for pending apps.")
```

File: pkg/annotation/annotation.go [hunk 4]
```diff
@@ -19,6 +19,10 @@ const (
 	// force is used when upgrading the Helm release.
 	ForceHelmUpgrade = "chart-operator.giantswarm.io/force-helm-upgrade"
 
+	// RollbackCount is the name of the annoataion storing the number of
```
Comment (rossf7) [inline pkg/annotation/annotation.go:22]:
```suggestion
	// RollbackCount is the name of the annotataion storing the number of
```

File: service/controller/chart/resource/release/update.go [hunk 5]
```diff
@@ -229,5 +245,96 @@ func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desir
...
+
+func (r *Resource) rollback(ctx context.Context, obj interface{}, currentStatus string) error {
+	cr, err := key.ToCustomResource(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	counts, ok := cr.GetAnnotations()[annotation.RollbackCount]
+	var rollbackCounts int
+	if !ok {
```
Comment (rossf7) [inline service/controller/chart/resource/release/update.go:264]:
```suggestion
	count, ok := cr.GetAnnotations()[annotation.RollbackCount]
	
	var rollbackCount int
	
	if !ok {
```

Can we add some newlines? Also minor but I think count should be singular to match the annotation.

File: service/controller/chart/resource/release/update.go [hunk 6]
```diff
@@ -229,5 +245,96 @@ func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desir
...
+		}
+	}
+
+	if rollbackCounts > r.maxRollback {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("the %#q release is in status %#q and cannot be updated despite of %d rollbacks", key.ReleaseName(cr), currentStatus, r.maxRollback))
+		return nil
+	}
+
+	if currentStatus == helmclient.StatusPendingInstall {
+		err = r.helmClient.DeleteRelease(ctx, key.Namespace(cr), key.ReleaseName(cr))
```
Comment (rossf7) [inline service/controller/chart/resource/release/update.go:279]: We should have logging before and after this. As deleting the release changes the state.

File: service/controller/chart/resource/release/update.go [hunk 7]
```diff
@@ -229,5 +245,96 @@ func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desir
...
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	} else {
+		// Rollback to revision 0 restore a release to the previous revision.
```
Comment (rossf7) [inline service/controller/chart/resource/release/update.go:284]: We should also have logging for the rollback.

File: pkg/annotation/annotation.go (hunk 4 above)
Comment (tomahawk28) [reply pkg/annotation/annotation.go:22]: I'm changing it to annotaion :) thanks

File: service/controller/chart/resource/release/update.go [hunk 8]
```diff
@@ -229,5 +245,106 @@ func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desir
...
 	return nil, nil
 }
+
+func (r *Resource) rollback(ctx context.Context, obj interface{}, currentStatus string) error {
+	cr, err := key.ToCustomResource(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	counts, ok := cr.GetAnnotations()[annotation.RollbackCount]
```
Comment (rossf7) [inline service/controller/chart/resource/release/update.go:262]:
```suggestion
	count, ok := cr.GetAnnotations()[annotation.RollbackCount]
```

I think better to use singular here too.

File: service/controller/chart/resource/release/update.go [hunk 9]
```diff
@@ -229,5 +245,106 @@ func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desir
...
+
+	var rollbackCount int
+
+	if !ok {
+		rollbackCount = 0
+	} else {
+		rollbackCount, err = strconv.Atoi(counts)
```
Comment (rossf7) [inline service/controller/chart/resource/release/update.go:269]:
```suggestion
		rollbackCount, err = strconv.Atoi(count)
```

File: service/controller/chart/resource/release/update.go [hunk 10]
```diff
@@ -229,5 +245,106 @@ func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desir
...
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	if rollbackCount > r.maxRollback {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("the %#q release is in status %#q and cannot be updated despite of %d rollbacks", key.ReleaseName(cr), currentStatus, r.maxRollback))
```
Comment (rossf7) [inline service/controller/chart/resource/release/update.go:276]:
```suggestion
		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("the %#q release is in status %#q and has reached %d rollbacks", key.ReleaseName(cr), currentStatus, r.maxRollback))
```

I tried to make this a bit shorter. WDYT?

File: service/controller/chart/resource/release/update.go (hunk 9 above)
Comment (tomahawk28) [reply service/controller/chart/resource/release/update.go:269]: oops, one more plural!

File: service/controller/chart/resource/release/update.go (hunk 10 above)
Comment (tomahawk28) [reply service/controller/chart/resource/release/update.go:276]: much better, thanks!

File: go.mod (hunk 1 above)
Comment (tomahawk28) [reply go.mod]: Done, back to `1.0.6`

**Comment by tomahawk28** [comment]
going to merge now. :)

**Pull request merged**